gips/data/sentinel2/sentinel2.py

```diff
@@ -0,0 +1,449 @@
+        `overwrite` is True, it will overwrite existing products if they
+        are found.  Products are saved to a well-known or else specified
+        directory.  kwargs is unused, and is present for compatibility.
+        """
+        self._time_report('Starting processing for this temporal-spatial unit')
+        products = self.needed_products(products, overwrite)
+        if len(products) == 0:
+            utils.verbose_out('No new processing required.')
+            return
+        for val in products.requested.values():
+            toa = (self._products[val[0]].get('toa', False) or 'toa' in val)
+            # for now users must use `-p` to get toa versions of products; getting all products in
+            # one go AND defaulting to toa doesn't seem to be supported.
+            if not toa:
+                utils.verbose_out('Please specify products with eg "-p ref-toa"', 1,
+                                  stream=sys.stderr)
+		raise NotImplementedError('only toa products are supported for now')
+
+        asset_type = 'L1C' # only one in the driver for now, conveniently
+
+        # construct as much of the product filename as we can right now
+        filename_prefix = os.path.join(
+                self.path, self.basename + '_' + self.sensors[asset_type] + '_')
+
+        # Read the assets
+        with utils.error_handler('Error reading '
+                                 + utils.basename(self.assets[asset_type].filename)):
+            self.read_raw() # returns a GeoImage, presently unused
+
+        md = self.meta_dict()
+
+        self._time_report('Starting upsample of Sentinel-2 asset bands')
+
+        sensor = self.sensors[asset_type]
+        # dict describing specification for all the bands in the asset
+        data_spec = self.assets[asset_type]._sensors[sensor]
+
+        # Make a proto-product which acts as a basis for several products, and is equivalent to
+        # ref-toa; it's needed because the asset's spatial resolution must be resampled to be equal
+        # for all bands of interest.
+        # TODO if the current date's ref-toa is already produced, open that instead, for performance
+        # compile a list of the files needed for the proto-product
+        # TODO this upsamples everything, when only two bands need it.  The first attempt to remedy
+        # this was actually slower, however; see branch 200-optimize-upsampling, commit 2e97ba0.
+        src_filenames = [f for f in self.metadata['abs-filenames']
+                if f[-6:-4] in data_spec['indices-bands']]
+        # upsample each one in turn (some don't need it but do them all for simplicity)
+        with utils.make_temp_dir() as tmpdir:
+            upsampled_filenames = [os.path.join(tmpdir, os.path.basename(f) + '.tif')
+                    for f in src_filenames]
+            for in_fn, out_fn in zip(src_filenames, upsampled_filenames):
+                cmd_str = 'gdalwarp -tr 10 10 {} {}'.format(in_fn, out_fn)
```

Comment:
I cannot recall at the moment how to constrain the processing here, but this command-line uses 4-cores when run on my laptop with GDAL2.  I think this is not a blocker to merging, but something that will irritate core-stingy users ;)